### PR TITLE
mgr/cephadm: set host crush location based on HostSpec

### PR DESCRIPTION
**Pull request by liewegas**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1400,6 +1400,17 @@ def _add_host(self, spec):
             raise OrchestratorError('New host %s (%s) failed check(s): %s' % (
                 spec.hostname, spec.addr, errors))
 
+        # prime crush map?
+        if spec.location:
+            ret, _, _ = self.mon_command({
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1405]:
```suggestion
            ret, _, _ = self.check_mon_command({
```

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -1400,6 +1400,17 @@ def _add_host(self, spec):
             raise OrchestratorError('New host %s (%s) failed check(s): %s' % (
                 spec.hostname, spec.addr, errors))
 
+        # prime crush map?
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1403]: should this be done everytime we save the host, not just once?

File: src/python-common/ceph/deployment/hostspec.py [hunk 3]
```diff
@@ -42,34 +43,55 @@ def __init__(self,
...
 
+        self.location = location
+
     def to_json(self) -> dict:
         return {
             'hostname': self.hostname,
             'addr': self.addr,
             'labels': list(OrderedDict.fromkeys((self.labels))),
             'status': self.status,
+            'location': self.location,
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/hostspec.py:54]: how about avoiding returning `None` here` this makes the yaml a bit more readable

**Comment by sebastian-philipp** [comment]
Can we remove the CRUSH map when using Ceph altogether? This seems far away, but still. By now, users no longer need to care about PGs. Things like this are great, as they make the learning curve for Ceph a more gentle. 

Which workflows still require manual CRUSH modifications? can we automate them as well?

File: src/pybind/mgr/cephadm/module.py (hunk 2 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/module.py:1403]: i think just once... i don't want to get into a situation where a user is trying to do ``ceph osd crush move ...`` and has to fight with cephadm.  (OTOH this means an export + import might move the item, but i think that's less bad.)

Comment (liewegas) [reply src/pybind/mgr/cephadm/module.py:1403]: actually, nm, ``add-bucket`` succeeds if the item exists in the wrong position, so i think it works fine as-is

**Comment by sebastian-philipp** [comment]
Uh, we have lots of bugs here. 

1. `ceph orch host add` is not idempotent. If you run 

    ```
    ceph orch apply -i <<EOF
    hostname: foobar
    ```
    we're calling `prime_empty_host`, which clears the hosts's inventory, potentially leading to deploying duplicated daemons on that host, if you're unlucky. 


2. Also, 

    https://github.com/ceph/ceph/blob/5f12e85d91543db5cd041ef2d840b8f4b98ce6be/src/pybind/mgr/cephadm/module.py#L1430-L1435

    completely evades the checks we have in `_add_host`m like `utils.resolve_ip(spec.addr)` or `check-host`.

**Review by jmolmo — APPROVED** [review]

**Pull request merged**